src/Dictionaries/DictionarySourceHelpers.cpp

```diff
@@ -130,4 +134,30 @@ String TransformWithAdditionalColumns::getName() const
 {
     return "TransformWithAdditionalColumns";
 }
+
+DictionaryPipelineExecutor::DictionaryPipelineExecutor(QueryPipeline & pipeline_, bool async)
+    : async_executor(async ? std::make_unique<PullingAsyncPipelineExecutor>(pipeline_) : nullptr)
+    , executor(async ? nullptr : std::make_unique<PullingPipelineExecutor>(pipeline_))
+{}
+
+bool DictionaryPipelineExecutor::pull(Block & block)
+{
+    if (async_executor)
+    {
+        while (true)
+        {
+            bool has_data = async_executor->pull(block);
+            if (has_data && !block)
+                continue;
+            return has_data;
+        }
+    }
+    else if (executor)
+        return executor->pull(block);
+    else
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "DictionaryPipelineExecutor is not initialized");
```

Comment:
In ctor it's clear that one of executors is initialized from initializer-list. I added this branch just in case if new ctor is added and there it may be forgotten.